src/github.com/getlantern/protected/protected.go

```diff
@@ -126,60 +130,89 @@ func Resolve(addr string) (*net.TCPAddr, error) {
+	op := ops.Begin("protected-dial").Set("origin", addr).Set("timeout", timeout.Seconds())
+	defer op.End()
+	conn, err := p.dial(op, network, addr, timeout)
+	return conn, op.FailIf(err)
+}
+
+func (p *Protector) dial(op ops.Op, network, addr string, timeout time.Duration) (net.Conn, error) {
+	start := time.Now()
+
+	remainingTimeout := func() time.Duration {
+		return timeout - time.Now().Sub(start)
+	}
+
 	host, port, err := SplitHostPort(addr)
 	if err != nil {
 		return nil, err
 	}
 
+	_ip, timedOut, err := withtimeout.Do(remainingTimeout(), func() (interface{}, error) {
+		// do DNS query
+		ip := net.ParseIP(host)
+		if ip != nil {
+			return ip, nil
+		}
+		// Try to resolve it
+		addr, err := p.Resolve(network, addr)
+		if err != nil {
+			return nil, err
+		}
+		return addr.IP, nil
+	})
+
+	if timedOut {
+		return nil, errors.New("Timed out looking up address %v within %v", addr, timeout)
+	} else if err != nil {
+		return nil, errors.New("Unable to look up address %v: %v", addr, err)
+	}
+
+	ip := _ip.(net.IP)
+
 	conn := &ProtectedConn{
 		port: port,
 	}
-	// do DNS query
-	IPAddr := net.ParseIP(host)
-	if IPAddr == nil {
-		log.Errorf("Couldn't parse IP address %v", host)
-		return nil, err
-	}
-
-	copy(conn.ip[:], IPAddr.To4())
+	copy(conn.ip[:], ip.To4())
 
 	socketFd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_STREAM, 0)
 	if err != nil {
-		log.Errorf("Could not create socket: %v", err)
-		return nil, err
+		return nil, errors.New("Could not create socket: %v", err)
 	}
 	conn.socketFd = socketFd
-
 	defer conn.cleanup()
 
-	// Actually protect the underlying socket here
-	protect, _ := getCurrent()
-	err = protect(conn.socketFd)
-	if err != nil {
-		return nil, fmt.Errorf("Could not bind socket to system device: %v", err)
+	_, timedOut, err = withtimeout.Do(remainingTimeout(), func() (interface{}, error) {
+		// Actually protect the underlying socket here
+		return nil, p.protect(conn.socketFd)
+	})
+	if timedOut {
+		return nil, errors.New("Timed out protecting socket to %v within %v", addr, timeout)
+	} else if err != nil {
+		return nil, errors.New("Unable to protect socket to %v: %v", addr, err)
 	}
 
-	err = conn.connectSocket()
-	if err != nil {
-		log.Errorf("Could not connect to socket: %v", err)
-		return nil, err
+	_, timedOut, err = withtimeout.Do(remainingTimeout(), func() (interface{}, error) {
+		// Actually protect the underlying socket here
+		return nil, conn.connectSocket()
+	})
+	if timedOut {
+		return nil, errors.New("Timed out connecting socket to %v within %v", addr, timeout)
+	} else if err != nil {
+		return nil, errors.New("Unable to connect socket to %v: %v", addr, err)
 	}
 
 	// finally, convert the socket fd to a net.Conn
 	err = conn.convert()
 	if err != nil {
-		log.Errorf("Error converting protected connection: %v", err)
-		return nil, err
+		return nil, errors.New("Error converting protected connection: %v", err)
 	}
-
-	conn.Conn.SetDeadline(time.Now().Add(timeout))
```

Comment:
This logic seemed wrong.  SetDeadline sets an absolute deadline on all operations on this connection, whereas the timeout is just intended to govern the dial.  So, I changed this method to use withtimeout to enforce the dial timeout just on the dial phase.